### PR TITLE
PR: Implemented Objects & WS Caching

### PR DESCRIPTION
**Pull request by hirohitonumberonefan**
**List of things added:**
- Implemented WS caching
- Implemented the objects made
- Added `ActionRow` object for ActionRows

### CODE REVIEW

File: src/DankCord/DankCord.py [hunk 1]
```diff
@@ -145,20 +145,23 @@ def run_command(self, /, name: str, retry=False, attempts=10, timeout: int = 10)
...
-    message_object = False
     for i in range(attempts):
-      requests.post("https://discord.com/api/v9/interactions", json.dumps(data), http_headers=[("Authorization", self.token), ("Content-Type", "application/json")])
+      response = requests.post("https://discord.com/api/v9/interactions", json.dumps(data), http_headers=[("Authorization", self.token), ("Content-Type", "application/json")])
       message_data = self._confirm_command_ran(nonce, timeout)
       status = not message_data is None
       if retry is False or status is True:
         break
     if status:
+      while not str(nonce) in self.ws_cache.keys():
```
Comment (ThePrivatePanda) [inline src/DankCord/DankCord.py]: calling `str()` so many times adds delay and load to cpu and mem, either store it as a variable outside the loop or better, confirm if it even requires converstion to string.

File: src/DankCord/DankCord.py [hunk 2]
```diff
@@ -145,20 +145,23 @@ def run_command(self, /, name: str, retry=False, attempts=10, timeout: int = 10)
...
+        time.sleep(0.00001)
```
Comment (ThePrivatePanda) [inline src/DankCord/DankCord.py]: Unnecessary explicit wait

File: src/DankCord/DankCord.py [hunk 3]
```diff
@@ -194,16 +197,20 @@ def run_sub_command(self, /, name: str, sub_name: str, retry=False, attempts=10,
...
-    message_object = False
     for i in range(attempts):
       requests.post("https://discord.com/api/v9/interactions", json.dumps(data), http_headers=[("Authorization", self.token), ("Content-Type", "application/json")])
       message_data = self._confirm_command_ran(nonce, timeout)
       status = not message_data is None
       if retry is False or status is True:
         break
     if status:
+      while not str(nonce) in self.ws_cache.keys():
+        time.sleep(0.00001)
```
Comment (ThePrivatePanda) [inline src/DankCord/DankCord.py]: Same as before

File: src/DankCord/DankCord.py [hunk 4]
```diff
@@ -68,17 +69,16 @@ def _confirm_command_ran(self, nonce_used: str, timeout: int):
...
 
-  def _events_listener_test(self):
-    return
+  def _events_listener(self):
     while True:
       event = json.loads(self.ws.recv())
       try:
-        if len(event["d"]["embeds"]) > 0:
-          print(event["d"]["embeds"])
+        if event["d"]["channel_id"] == str(self.channel_id) or event["d"]["channel_id"] == "270904126974590976":
```
Comment (theAbdoSabbagh) [inline src/DankCord/DankCord.py:80]: Why is the channel ID hardcoded here?

File: src/DankCord/Objects.py [hunk 5]
```diff
@@ -33,21 +45,46 @@ def __init__(self, data : dict) -> None:
...
+    except:
+      self.emoi = None
+    try:
+      self.label : str = data["label"]
+    except:
+      self.label : str = ""
+    try:
+      self.disabled : bool = data["disabled"]
+    except:
+      self.disabled : bool = False
```
Comment (ThePrivatePanda) [inline src/DankCord/Objects.py:87]: this part can be cleaned up a little with conditional and oneline statements

File: src/DankCord/Objects.py [hunk 6]
```diff
@@ -66,11 +103,23 @@ def __init__(self, data : dict) -> None:
...
+      self.description : str = ""
+    try:
+      self.url : str = data["url"]
+    except:
+      self.url : str = ""
+    try:
+      self.footer : EmbedFooter = EmbedFooter(data["footer"])
+    except:
+      self.footer : EmbedFooter = None
+
```
Comment (ThePrivatePanda) [inline src/DankCord/Objects.py:122]: same as before, can be cleaned up

**Review by ThePrivatePanda — CHANGES_REQUESTED** [review]

File: src/DankCord/DankCord.py (hunk 2 above)
Comment (theAbdoSabbagh) [inline src/DankCord/DankCord.py]: Does sleep really matter here?

File: src/DankCord/DankCord.py (hunk 3 above)
Comment (theAbdoSabbagh) [inline src/DankCord/DankCord.py]: Why time.sleep instead of just continue?

File: src/DankCord/DankCord.py [hunk 7]
```diff
@@ -145,20 +145,23 @@ def run_command(self, /, name: str, retry=False, attempts=10, timeout: int = 10)
...
       if retry is False or status is True:
         break
     if status:
+      while not str(nonce) in self.ws_cache.keys():
+        time.sleep(0.00001)
+      message_data = self.ws_cache[str(nonce)]
       message_object = Message(message_data)
-    return message_object
+      return message_object
+    else:
```
Comment (theAbdoSabbagh) [inline src/DankCord/DankCord.py]: No need for if and else here, u can just add return none in the line after message_object with 1 less indent

File: src/DankCord/DankCord.py [hunk 8]
```diff
@@ -194,16 +197,20 @@ def run_sub_command(self, /, name: str, sub_name: str, retry=False, attempts=10,
...
```
Comment (theAbdoSabbagh) [inline src/DankCord/DankCord.py]: Again, no need for if and else here, u can just add return none in the line after message_object with 1 less indent

File: src/DankCord/DankCord.py (hunk 3 above)
Comment (hirohitonumberonefan) [reply src/DankCord/DankCord.py]: Changed

File: src/DankCord/DankCord.py (hunk 2 above)
Comment (hirohitonumberonefan) [reply src/DankCord/DankCord.py]: Changed

File: src/DankCord/DankCord.py (hunk 3 above)
Comment (hirohitonumberonefan) [reply src/DankCord/DankCord.py]: Changed

File: src/DankCord/DankCord.py (hunk 1 above)
Comment (hirohitonumberonefan) [reply src/DankCord/DankCord.py]: changed

File: src/DankCord/Objects.py [hunk 9]
```diff
@@ -33,21 +45,46 @@ def __init__(self, data : dict) -> None:
...
 
 class Button:
   def __init__(self, data : dict) -> None:
-    self.emoji : Emoji = Emoji(data["emoji"])
-    self.disabled : bool = data["disabled"]
+    self.type = 2
+    try:
+      self.emoji : Emoji = Emoji(data["emoji"])
+    except:
+      self.emoi = None
```
Comment (theAbdoSabbagh) [inline src/DankCord/Objects.py]: typo

File: src/DankCord/DankCord.py (hunk 4 above)
Comment (hirohitonumberonefan) [reply src/DankCord/DankCord.py:80]: oh thats the dank memer channel

File: src/DankCord/DankCord.py (hunk 2 above)
Comment (hirohitonumberonefan) [reply src/DankCord/DankCord.py]: changed

File: src/DankCord/DankCord.py (hunk 7 above)
Comment (hirohitonumberonefan) [reply src/DankCord/DankCord.py]: Done

File: src/DankCord/Objects.py (hunk 5 above)
Comment (ThePrivatePanda) [reply src/DankCord/Objects.py:87]: delayed

File: src/DankCord/Objects.py (hunk 6 above)
Comment (ThePrivatePanda) [reply src/DankCord/Objects.py:122]: delayed

File: src/DankCord/Objects.py (hunk 9 above)
Comment (hirohitonumberonefan) [reply src/DankCord/Objects.py]: fixed

File: src/DankCord/DankCord.py (hunk 8 above)
Comment (hirohitonumberonefan) [reply src/DankCord/DankCord.py]: changed

**Review by ThePrivatePanda — APPROVED** [review]

**Pull request merged**